File: MinJu/231204_week/Programmers_12978_배달.py
from heapq import heappop, heappush

def solution(N, road, K):
    answer = 0
    distance = [1e9] * (N+1)  # 각 마을까지의 최단 거리를 저장, 초기값은 무한대 
    graph = [[] for _ in range(N+1)]  # 각 마을과 연결된 도로 정보를 담은 인접 리스트

    # 그래프 정보 설정
    for a, b, c in road:
        graph[a].append((b, c))
        graph[b].append((a, c))

    distance[1] = 0  # 자신 -> 자신의 거리는 0

    # 우선순위 큐를 사용한 다익스트라 알고리즘
    min_heap = [(0, 1)]  # (거리, 노드)

    while min_heap:
        dist, node = heappop(min_heap)

        if distance[node] < dist:
            continue  # 이미 더 짧은 경로가 들어가있다면 건너뛰기 

        for next_node, next_dist in graph[node]:
            new_dist = dist + next_dist
            if distance[next_node] > new_dist:
                distance[next_node] = new_dist
                heappush(min_heap, (new_dist, next_node))  # 더 짧은 거리이면 우선순위 큐에 추가

    #  K 이하인 거리의 마을 개수 세기
    for new_dist in distance[1:]:
        if new_dist <= K:
            answer += 1

    return answer

File: SooBin/231204_week/Programmers_12978_배달.cpp
#include <iostream>
#include <vector>

#define INF 1e9

using namespace std;

int graph[51][51];  // 마을의 개수는 1이상 50이하의 자연수

int solution(int N, vector<vector<int> > road, int K) {
    int answer = 0;
    
    // 최단 거리 테이블을 모두 무한으로 초기화
    for(int i = 0; i < 51; i++){
        fill(graph[i],graph[i]+51,INF);
    }
    
    // 자기 자신에서 자기 자신으로 가는 비용은 0으로 초기화
    for(int a = 1; a <= N; a++){
        for(int b = 1; b <= N; b++){
            if(a == b) graph[a][b] = 0;
        }
    }
    
    for(int i = 0; i < road.size();i++){
        // 양방향 도로.
        // 한 도로의 정보가 중복해서 주어질 수 있으므로 입력값 중 최소값으로 갱신한다.
        graph[road[i][0]][road[i][1]] = min(road[i][2],graph[road[i][0]][road[i][1]]);
        graph[road[i][1]][road[i][0]] = min(road[i][2],graph[road[i][0]][road[i][1]]);
    }
    
    
    // 플로이드 워셜 알고리즘 수행
    for(int k = 1; k <= N; k++) {
        for(int a = 1; a <= N; a++){
            for(int b = 1; b <= N; b++){
                graph[a][b] = min(graph[a][b],graph[a][k] + graph[k][b]);
            }
        }
    }
    
    for(int i = 1; i <= N; i++){
        if(graph[1][i] <= K){
            answer++;
        }
    }
     

    cout << answer << endl;

    return answer;
}